aws/resource_aws_docdb_cluster.go

```diff
@@ -0,0 +1,796 @@
+	d.Set("storage_encrypted", dbc.StorageEncrypted)
+
+	var vpcg []string
+	for _, g := range dbc.VpcSecurityGroups {
+		vpcg = append(vpcg, aws.StringValue(g.VpcSecurityGroupId))
+	}
+	if err := d.Set("vpc_security_group_ids", vpcg); err != nil {
+		return fmt.Errorf("error setting vpc_security_group_ids: %s", err)
+	}
+
+	// Fetch and save tags
+	if err := saveTagsDocDB(conn, d, aws.StringValue(dbc.DBClusterArn)); err != nil {
+		log.Printf("[WARN] Failed to save tags for DocDB Cluster (%s): %s", aws.StringValue(dbc.DBClusterIdentifier), err)
+	}
+
+	return nil
+}
+
+func resourceAwsDocDBClusterUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).docdbconn
+	requestUpdate := false
+
+	req := &docdb.ModifyDBClusterInput{
+		ApplyImmediately:    aws.Bool(d.Get("apply_immediately").(bool)),
+		DBClusterIdentifier: aws.String(d.Id()),
+	}
+
+	if d.HasChange("master_password") {
+		req.MasterUserPassword = aws.String(d.Get("master_password").(string))
+		requestUpdate = true
+	}
+
+	if d.HasChange("engine_version") {
+		req.EngineVersion = aws.String(d.Get("engine_version").(string))
+		requestUpdate = true
+	}
+
+	if d.HasChange("vpc_security_group_ids") {
+		if attr := d.Get("vpc_security_group_ids").(*schema.Set); attr.Len() > 0 {
+			req.VpcSecurityGroupIds = expandStringList(attr.List())
+		} else {
+			req.VpcSecurityGroupIds = []*string{}
+		}
+		requestUpdate = true
+	}
+
+	if d.HasChange("preferred_backup_window") {
+		req.PreferredBackupWindow = aws.String(d.Get("preferred_backup_window").(string))
+		requestUpdate = true
+	}
+
+	if d.HasChange("preferred_maintenance_window") {
+		req.PreferredMaintenanceWindow = aws.String(d.Get("preferred_maintenance_window").(string))
+		requestUpdate = true
+	}
+
+	if d.HasChange("backup_retention_period") {
+		req.BackupRetentionPeriod = aws.Int64(int64(d.Get("backup_retention_period").(int)))
+		requestUpdate = true
+	}
+
+	if d.HasChange("db_cluster_parameter_group_name") {
+		d.SetPartial("db_cluster_parameter_group_name")
+		req.DBClusterParameterGroupName = aws.String(d.Get("db_cluster_parameter_group_name").(string))
+		requestUpdate = true
+	}
+
+	if d.HasChange("enabled_cloudwatch_logs_exports") {
+		d.SetPartial("enabled_cloudwatch_logs_exports")
+		req.CloudwatchLogsExportConfiguration = buildDocDBCloudwatchLogsExportConfiguration(d)
+		requestUpdate = true
+	}
+
+	if requestUpdate {
+		err := resource.Retry(5*time.Minute, func() *resource.RetryError {
+			_, err := conn.ModifyDBCluster(req)
+			if err != nil {
+				if isAWSErr(err, "InvalidParameterValue", "IAM role ARN value is invalid or does not include the required permissions") {
+					return resource.RetryableError(err)
+				}
+
+				if isAWSErr(err, docdb.ErrCodeInvalidDBClusterStateFault, "Cannot modify engine version without a primary instance in DB cluster") {
+					return resource.NonRetryableError(err)
+				}
+
+				if isAWSErr(err, docdb.ErrCodeInvalidDBClusterStateFault, "") {
```

Comment:
Was this copied from RDS/Neptune? We should strongly prefer to only handle certain error messages instead of blanket retrying on all error messages with this error code. I'm guessing what we really want is similar to what is present in the delete function and we can remove the conditional above:

```go
if isAWSErr(err, docdb.ErrCodeInvalidDBClusterStateFault, "is not currently in the available state") {
  return resource.RetryableError(err)
}
```

If this was copied and we are really unsure what it should be, we should remove this conditional and the one above it so we can correctly implement the error retry conditions (if/when they appear in GitHub issues).

Reply:
I copied most of the code from `aws_rds_cluster`.

I've just tried to remove both of the conditionals, but the acceptance tests failed with:
```
--- FAIL: TestAccAWSDocDBCluster_updateCloudwatchLogsExports (183.52s)
    testing.go:538: Step 1 error: Error applying: 1 error occurred:
                * aws_docdb_cluster.default: 1 error occurred:
                * aws_docdb_cluster.default: Failed to modify DocDB Cluster (tf-docdb-cluster-2244346464682253664): InvalidDBClusterStateFault: DB cluster is not available for modification with status backing-up
                status code: 400, request id: 0df2055c-efed-4a70-a4a9-5004bfadbe4b
```

So I removed the two conditionals and replaced them with:
```
				if isAWSErr(err, docdb.ErrCodeInvalidDBClusterStateFault, "DB cluster is not available for modification") {
					return resource.RetryableError(err)
				}

```